### PR TITLE
Create Python Script to download, extract YouTube videos and upload it to S3

### PR DESCRIPTION
**Pull request by AjayRangarajan**
- [x] Add functions to download, extract and upload videos to S3.
> Removed multi threading from the script since it caused **RuntimeError** when tried to upload video to S3.
>
> Will add multi threading in the next commit.

### CODE REVIEW

File: scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+### Install packages
+
+Using pipenv:
+```
+pipenv install
+```
+
+Using venv:
```
Comment (sutyum) [inline scripts/README.md:8]: not needed

File: scripts/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+boto3==1.26.8
```
Comment (sutyum) [inline scripts/requirements.txt:1]: this file is not needed

File: scripts/script.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+# from threading import Thread
+import re
+import os
+import argparse
+import sys
+import boto3
+from botocore.exceptions import ClientError
+from dotenv import load_dotenv
+
+
```
Comment (sutyum) [inline scripts/script.py:13]: better to move everything in the script portion inside an if main block:
```python
if __name__ == '__main__':
  ...
```

File: scripts/script.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+EXTRACTED_ADS_PATH = fr"{parsed_args.extracted_ads_path}"
+S3_BUCKET_NAME = fr"{parsed_args.s3_bucket_name}"
+
+
+
+def download_yt_video(video_url, timestamps):
+    """Download Youtube Videos and Extract into subclips
+
+    Args:
+        video_url (str): The URL of the Youtube Video
```
Comment (sutyum) [inline scripts/script.py:36]: would be better to add type annotation to the function signature too

File: scripts/script.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        timestamps (list[str]): List of timestamps used to extract the video into subclips
+    """
+    yt = YouTube(video_url)
+    yt.streams.filter(file_extension='mp4')
+    vid = yt.streams.get_by_itag(22)
```
Comment (sutyum) [inline scripts/script.py:41]: what does `22` signify here?

File: scripts/script.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import boto3
+from botocore.exceptions import ClientError
+from dotenv import load_dotenv
+
+
+load_dotenv()
+
+
+parser = argparse.ArgumentParser(description='Download Youtube video, extract into subclips and upload to S3')
+parser.add_argument('video_list', help='Video list location')
```
Comment (sutyum) [inline scripts/script.py:18]: do you think click maybe a"nicer" option here?

File: scripts/script.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    Args:
+        video_url (str): The URL of the Youtube Video
+        timestamps (list[str]): List of timestamps used to extract the video into subclips
+    """
+    yt = YouTube(video_url)
+    yt.streams.filter(file_extension='mp4')
+    vid = yt.streams.get_by_itag(22)
+    vid_title = vid.title
+    print(f"Started downloading video: {vid_title}")
+    downloaded_video_path = vid.download(DOWNLOAD_PATH)
```
Comment (sutyum) [inline scripts/script.py:44]: would be nice to see a progress bar for this process. Would it be useful to use `tqdm`?

File: scripts/script.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        extract_subclip(downloaded_video_path, timestamp)
+
+def extract_subclip(file_location, timestamp):
+    """Extract videos into subclips
+
+    Args:
+        file_location (str): Location of video file that needs to be extracted into subclips
+        timestamp (str): timestamp of the video which needs to be extracted into a subclip
+    """
+    start_min, start_sec, end_min, end_sec = map(int, re.split('[:-]', timestamp))
```
Comment (sutyum) [inline scripts/script.py:57]: you should do some checks here to ensure that the entire csv file is valid, and return reasonable errors if the csv file does not have the expected format

Comment (sutyum) [reply scripts/script.py:57]: You could potentially use pydantic for validating the format of all the timestamps

File: scripts/script.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+AWS_ACCESS_KEY = os.environ.get('AWS_ACCESS_KEY')
+AWS_SECRET_ACCESS_KEY = os.environ.get('AWS_SECRET_ACCESS_KEY')
+
+s3_client = boto3.client(
+    's3',
+    aws_access_key_id=AWS_ACCESS_KEY,
+    aws_secret_access_key=AWS_SECRET_ACCESS_KEY
+)
+def upload_to_s3(file_location, target_filename):
```
Comment (sutyum) [inline scripts/script.py:76]: dead code?

File: scripts/script.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    """Upload the video to S3
+
+    Args:
+        file_location (str): Location of the Video that needs to be uploaded to S3.
+        target_filename (str): Name of the uploaded file in S3
+
+    Returns:
+        bool: returns True if the video is successfully uploaded else, returns False
+    """
+    try:
```
Comment (sutyum) [inline scripts/script.py:86]: this function is an empty abstraction, why do you need this?

File: scripts/script.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    except ClientError as e:
+        print(e)
+        return False
+    except Exception as e:
+        print(e)
+        return False
+    return True
+
+
+with open(VIDEO_LIST, mode ='r')as file:
```
Comment (sutyum) [inline scripts/script.py:97]: script bits should be in main block

**Review by sutyum — CHANGES_REQUESTED** [review]
The commit is over all quite a good first commit. While various changes suggested are a matter of style, there is one key fundamental problem.

the flow of script should broadly follow the following procedure:
- open csv file
- parse the the csv file. if there are any issue with the format, throw reasonable errors. the output of this step should be a well constucted python object (Read: [Parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)
- carry out downloads (show progress bars while the user waits: CLIs should never seem like they are stuck)
- create subclips (progress bar needed?) [what happens when the timestamp is outside the range of the length of the video?]
- upload (progress bar needed?)
- Report: Number of clips downloaded, number of subclass created, number of subclass uploaded, number of subclass failed to create or upload etc

File: scripts/script.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    AWS_ACCESS_KEY = os.environ.get('AWS_ACCESS_KEY')
+    AWS_SECRET_ACCESS_KEY = os.environ.get('AWS_SECRET_ACCESS_KEY')
+    s3_client = boto3.client(
+        's3',
+        aws_access_key_id=AWS_ACCESS_KEY,
+        aws_secret_access_key=AWS_SECRET_ACCESS_KEY
+    )
+
+    with open(VIDEO_LIST, mode ='r')as file:
+        csvFile = csv.reader(file)
```
Comment (sutyum) [inline scripts/script.py:119]: what if this file doesn't exist?

File: scripts/script.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    parser.add_argument('s3_bucket_name', help='Name of the AWS S3 Bucket where downloaded Ads will be uploaded.')
+    parsed_args = parser.parse_args(sys.argv[1:])
+
+
+    VIDEO_LIST = fr"{parsed_args.video_list}"
+    S3_BUCKET_NAME = fr"{parsed_args.s3_bucket_name}"
+
+    download_folder = "YTVids"
+    if not os.path.exists(download_folder):
+        os.mkdir(download_folder)
```
Comment (sutyum) [inline scripts/script.py:105]: this may fail: if this fails: no need to do anything

File: scripts/script.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        's3',
+        aws_access_key_id=AWS_ACCESS_KEY,
+        aws_secret_access_key=AWS_SECRET_ACCESS_KEY
+    )
+
+    with open(VIDEO_LIST, mode ='r')as file:
+        csvFile = csv.reader(file)
+        for row in csvFile:
+            yt_url = row[0]
+            timestamps = row[1].split(';')
```
Comment (sutyum) [inline scripts/script.py:122]: parsing all over the codebase

**Comment by sutyum** [comment]
You need to resolve the conversations to proceed on this PR

File: scripts/README.md (hunk 1 above)
Comment (AjayRangarajan) [reply scripts/README.md]: Removed in the latest commit

File: scripts/requirements.txt (hunk 2 above)
Comment (AjayRangarajan) [reply scripts/requirements.txt]: Removed and added Pipfile in the next commit

File: scripts/script.py (hunk 3 above)
Comment (AjayRangarajan) [reply scripts/script.py]: Moved the code to the if block as mentioned.

File: scripts/script.py (hunk 4 above)
Comment (AjayRangarajan) [reply scripts/script.py]: Add type annotations in the next commit

File: scripts/script.py (hunk 7 above)
Comment (AjayRangarajan) [reply scripts/script.py]: Individual progress bar is not showing up when downloading videos in multi thread. When downloading as Thread pools, a single progress bar appears for the complete download process.

File: scripts/script.py (hunk 8 above)
Comment (AjayRangarajan) [reply scripts/script.py]: Implemented custom validator for parsing the csv file and validating.

File: scripts/script.py (hunk 9 above)
Comment (AjayRangarajan) [reply scripts/script.py]: Removed in the next commit.

File: scripts/script.py (hunk 10 above)
Comment (AjayRangarajan) [reply scripts/script.py:150]: Removed in the next commit

File: scripts/script.py (hunk 11 above)
Comment (AjayRangarajan) [reply scripts/script.py]: Moved to the main block as mentioned.

File: scripts/script.py (hunk 12 above)
Comment (AjayRangarajan) [reply scripts/script.py]: Added a condition to check if the file exists in the next commit.

File: scripts/script.py (hunk 14 above)
Comment (AjayRangarajan) [reply scripts/script.py]: Created a custom validator which will parse the file and take care of all validation. Which will also return a pydantic object that will be used in the download and extract functions.

File: scripts/.gitignore [hunk 15]
```diff
@@ -0,0 +1,172 @@
+# references
```
Comment (sutyum) [inline scripts/.gitignore:1]: move the folder into a subdirectory of scripts named aptly: scripts/s3_uploader_cli/

File: scripts/README.md [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+
+Using pipenv:
+```
+pipenv install
+```
+
+### Run the script:
+
+```
+python script.py <video_list.csv> <s3_bucket_name>
```
Comment (sutyum) [inline scripts/README.md:11]:
```
piping shell
python ...
```

File: scripts/script.py [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+            return False
+
+    def parse(self):
+        with open(self.video_list, mode ='r') as file:
+            video_list = []
+            csvFile = csv.reader(file)
+            csvList = list(enumerate(csvFile))
+            length = len(csvList)
+            if not length:
+                raise Exception(f"The csv file {self.video_list} is empty!")
```
Comment (sutyum) [inline scripts/script.py:60]: `ValueError()`

Comment (sutyum) [reply scripts/script.py:60]:
```
class EmptyCsvFileError(Exception):
  pass
```

File: scripts/script.py [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+from dotenv import load_dotenv
+
+
+load_dotenv()
+
+
+download_threads = []
+upload_threads = []
+
+class TimeStamp(BaseModel):
```
Comment (sutyum) [inline scripts/script.py:26]: the validator for timestamps would go here

File: scripts/script.py [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+    start_min: int
+    start_sec: int
+    end_min: int
+    end_sec: int
+
+class VideoDetail(BaseModel):
```
Comment (sutyum) [inline scripts/script.py:32]: validator for url would go here

File: scripts/script.py [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+            else:
+                return False
+        except:
+            return False
+
+    def parse(self):
+        with open(self.video_list, mode ='r') as file:
+            video_list = []
+            csvFile = csv.reader(file)
+            csvList = list(enumerate(csvFile))
```
Comment (sutyum) [inline scripts/script.py:57]: `video_list = VideoList(list(enumerate(csvfile)))` --> python object with a the data. The key point is that all our data in this object are guaranteed to be in the correct format.

File: scripts/script.py [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+    start_sec: int
+    end_min: int
+    end_sec: int
+
+class VideoDetail(BaseModel):
+    url: str
+    timestamps: List[TimeStamp]
+
+class VideoList(BaseModel):
+    entry: List[VideoDetail]
```
Comment (sutyum) [inline scripts/script.py:37]: can line count validation go in here or it has to be in the parse function?

File: scripts/script.py [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+        yt.streams.filter(file_extension='mp4')
+        vid = yt.streams.get_by_itag(22)
+        vid_title = vid.title
+        print(f"Started downloading video: {vid_title}")
+        downloaded_video_path = vid.download(DOWNLOAD_PATH)
+        print(f"Successfully downloaded video: {vid_title}\nDownload path: {downloaded_video_path}")
+        print(f"Extracting subclips of the video: {vid_title}")
+        for timestamp in vd.timestamps:
+            extract_subclip(downloaded_video_path, timestamp)
+    except Exception as e:
```
Comment (sutyum) [inline scripts/script.py:112]: you only need to catch exception if you are going to act on it. If it is an error that will be propagated to the CLI user, no need to catch

Comment (sutyum) [inline scripts/script.py:112]: `except Exception as e:` -> this is a major anti pattern

File: scripts/script.py [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+    """Extract videos into subclips
+
+    Args:
+        file_location (str): Location of video file that needs to be extracted into subclips
+        timestamp (TimeStamp): TimeStamp object that contains start_min, start_sec, end_min, end_sec attributes
+    """
+    starttime = (timestamp.start_min * 60) + timestamp.start_sec
+    endtime = (timestamp.end_min * 60) + timestamp.end_sec
+    _, tail = os.path.split(file_location)
+    filename = tail[0:-5]
```
Comment (sutyum) [inline scripts/script.py:126]: why this needed?

**Review by sutyum — CHANGES_REQUESTED** [review]

**Review by sutyum — APPROVED** [review]

**Review by sutyum — APPROVED** [review]

**Review by sutyum — APPROVED** [review]

**Pull request merged**